File: GD.R
# Title     : TODO
# Objective : TODO
# Created by: tinan
# Created on: 2021-04-26
p <- 18/37
xGrid <- 1:7
N <- 10^6
mcC <- c()
rouletteSpins <- function (p){
  x <- 0
  while (TRUE){
    x <- x + 1
    if (runif(1) < p) {
      return (x)
    }
  }
}
for (i in 1:N){
  mcC<- append(mcC, rouletteSpins(p))
}
mcC <- table(mcC)
mcEstimate  <- mcC/N
mcEstimate

gDist <- dgeom(xGrid-1,p)
gDist

plot(xGrid,mcEstimate[xGrid],type = 'h',col = 'blue', lwd = 10,
     xlab = 'X',ylab = 'Probability',main = 'A geometric PMF')
axis(2,at=seq(0,1,0.1))
points(gDist,col = 'red',lwd = 5,pch = 4)

legend("topright",legend = c("mcEstimate","PMF"),lty = c(1,2),
       col = c("blue","red"))

File: GD.py
import math
import random
import numpy as np
from matplotlib import pyplot as plt
from numpy.lib import append
import pandas as pd
def rouletteSpins(p: float):
    x = 0
    while True:
        x = x+1
        if random.random() < p:
            return x
            break

p, xGrid, N = 18/37, np.arange(1, 8), int(math.pow(10, 6))
mcC = []
for i in range(N):
    mcC = append(mcC, rouletteSpins(p))
mcEstimate = pd.value_counts(mcC)/N
print(mcEstimate)
gDist = np.random.geometric(p, N)
gPmf = pd.value_counts(gDist)/N
print(gPmf)

plt.bar(xGrid, mcEstimate[xGrid],width=0.3, color="blue", alpha=0.5)
plt.bar(xGrid, gPmf[xGrid], width=0.15, color='red', alpha=0.5)
plt.scatter(xGrid, mcEstimate[xGrid], color='blue', alpha=0.7)
plt.scatter(xGrid, gPmf[xGrid], color='red', alpha=0.7)
plt.xlim(0, 8)
plt.ylim(0, 0.5)
plt.yticks(np.arange(0, 1, 0.1))
plt.xlabel('X')
plt.ylabel('Probability')
plt.legend(["mcEstimate", "PMF"], loc='best', frameon=True)
plt.show()

